### ISSUE TITLE
Please comply with ASM license when re-distributing ASM

### ISSUE THREAD

**Issue by vlsi** [raphw/byte-buddy#709]
ASM license specifies that `Redistributions of source code must retain the ... copyright notice`

I see that `byte-buddy-1.10.0.jar` bundles ASM under `byte-buddy-1.10.0.jar/uzip://net/bytebuddy/jar` path, however it fails to retain the relevant copyright notice.

Would you please comply with ASM license? (see https://gitlab.ow2.org/asm/asm/blob/e246de13a995837bc88777aca47f141d9bcbfc01/LICENSE.txt )

Thanks.

PS. Here's relevant ASM issue: https://gitlab.ow2.org/asm/asm/issues/317880

**Comment by raphw** [comment]
Absolutely, this I really want to get right. 

The shaded source code already retains the BSD-3 license information in the source file headers. But there is currently no trace in the binary distribution. However, as point (2) it says:

> Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.

The BSD licensing of the shaded ASM dependency is mentioned in the Byte Buddy documentation and I am wondering if this is sufficient given the *and/or* bit.

Finally, do you know if there is a way to specify this sublicensing in the POM file of the *byte-buddy* artifact? If I specified the BSD 3 license there would this not imply that Byte Buddy is dual licensed?

**Comment by vlsi** [comment]
@raphw , frankly speaking, I would consider `pom.xml` contents as the "last resort".

Typically you have a jar, and you wonder "what is the license terms for that package".
It would be so much better if the conditions can be derived from the file itself rather than from some kind of manual lookup of the `pom.xml` or whatever files.

Note: it is great ByteBuddy has machine-readable `Bundle-License` manifest attribute.

I think the way to go is to include licensing information under `META-INF/LICENSE`.
In other words, `META-INF/LICENSE` could include default Apache-2.0 text for ByteBuddy itself followed by the relevant entries for the bundled dependencies.

Unfortunately, there's no "common format" for the licenses, so there's not much I can recommend there.

However there are two approaches:
1) You just append relevant bits to `META-INF/LICENSE`. For instance: https://github.com/apache/calcite-avatica/blob/8138d42841a599397fa6bedcb53ae694d388f100/LICENSE#L210-L242

As you have just a single bundled dependency, this seems to be the way to go.

2) You add a note like

```
Additional License files can be found in the 'licenses' folder located in the same directory as the LICENSE file (i.e. this file)

BSD-3-Clause
* org.ow2.asm:asm:7.1
```
and you include ASM license under `META-INF/license/...`

This approach seems to be better (easier to follow) when lots of dependencies are present.

**Comment by vlsi** [comment]
> The BSD licensing of the shaded ASM dependency is mentioned in the Byte Buddy documentation and I am wondering if this is sufficient given the and/or bit.

Which documentation do you mean?
Can you please clarify where you have a statement that certain ByteBuddy binaries include ASM
?

First of all, you need to identify what do you mean by "release".
For instance: https://bytebuddy.net/#/, "Getting Byte Buddy".

>Byte Buddy is currently released in version 1.10.0. You can download Byte Buddy directly from its Bintray page

Neither [bintray page](https://bintray.com/raphw/maven/ByteBuddy) nor relevant bytebuddy.jar file includes the required ASM copyright.

**Comment by vlsi** [comment]
@raphw , here's one more issue:

>https://bytebuddy.net/#/
>Please note that Byte Buddy depends on [the ASM library which is distributed under a BSD license](http://asm.ow2.org/license.html).

There are three issues there:
1) The URL is invalid
2) You should avoid wording like "BSD" because there are lots of BSD-like licenses. The better wording would be `BSD-3-Clause`
3) The ASM license might vary over time, so it is not that safe to have a single URL.

**Comment by vlsi** [comment]
>The BSD licensing of the shaded ASM dependency is mentioned in the Byte Buddy documentation and I am wondering if this is sufficient given the and/or bit.

That would work if the only artifact you released was `byte-buddy.zip` that contained the relevant readme and licensing information.

However you do release individual jar files that are loosely tied to pom.xml, so I would keep "pom licensing for ByteBuddy, and avoid mentioning ASM license there".

**Comment by raphw** [comment]
Alright, beginning with the next release, Byte Buddy includes a `LICENSE` and a `NOTICE` file in every jar that is distributed. This is more or less also what the Apache 2.0 license suggests.

As for the shading of ASM, I found this segment that suggests that one should reference the BSD-3 clause as it is compatible with Apache 2.0. The text I include in the *byte-buddy* artifact will be:

> This product bundles ASM ${version.asm}, which is available under a "3-clause BSD" license. For details, see ${asm.url}/license.html.

I got this from [Apache's own summary for licensing](http://www.apache.org/dev/licensing-howto.html).

**Comment by vlsi** [comment]
> It says that it is sufficient to add a pointer to a permissive license such as BSD-3.

Apache summary:

>In LICENSE, add a pointer to the dependency's license **within** the **distribution** and a short note summarizing its licensing

You can't use a pointer that points **outside** of your binary.

**Comment by vlsi** [comment]
>For details, see ${asm.url}/license.html.

Note:
1) You can't rely on `${asm.url}/license.html` being accessible for everyone for infinite amount of time
2) You can't even know if the "current page" under your lists the license for exact ASM version you bundle

The above implies one should never try to use "license url" as the only source of the license, especially when the license is cusomized (e.g. includes non-standart items like copyright notice)

**Comment by raphw** [comment]
That's right. I will copy the license information out of the shaded ASM jar once this is available and include it manually until then.

**Comment by vlsi** [comment]
> I will copy the license information out of the shaded ASM jar once this is available

Now's your turn to convince ASM developers to include license to the distributed jar files :)

**Comment by raphw** [comment]
I see the grief. Getting this to work was much more difficult then it was supposed to.

I wish such license bundling was just built-into Maven or the maven-license-plugin but I had to jump through quite a bunch of hoops.

I will propbably release a new version tonight and hopefullt get the edge problems sorted out until then.

**Comment by vlsi** [comment]
>I see the grief. Getting this to work was much more difficult then it was supposed to.

Just in case: Gradle makes things easier.

**Comment by raphw** [comment]
This is now fixed on master and will be part of the next release.

**Comment by ebruneton** [comment]
About "the documentation and/or other materials provided with the distribution.", I'm not sure a .jar qualifies as "other materials provided with the distribution". My understanding is that the license should be easy to find and read for the end user, not buried into some binary file. It can of course be included in the .jar, but I'm not sure this is sufficient to comply with the BSD license (*).

This BSD clause dates back from 1990, when software was mostly distributed in physical boxes, CDs or DVDs. I'm not sure how it should be interpreted now, when software is downloaded from a source web site. If the source web site is "other materials provided with the distribution", then showing the ASM license there should be sufficient (as you do on http://bytebuddy.net).

A lawyer would know better, but I don't know who to ask.

(*) for me the .jar is the binary form (as opposed to the source form), and the BSD clause does not say that the license must be retained in the binary form, but in the other material. Presumably because the binary form can't be read by the end user.

**Comment by vlsi** [comment]
@ebruneton , 

>(*) for me the .jar is the binary form (as opposed to the source form), and the BSD clause does not say that the license must be retained in the binary form, but in the other material. Presumably because the binary form can't be read by the end user.

1) .jar is the only form that is distributed by byte-buddy

2) META-INF is a common folder for putting meta information

3) The second clause reads

> 2. **Redistributions** in binary form **must reproduce** the above copyright
>    notice, this list of conditions and the following disclaimer in the
>    documentation and/or other materials **provided with the distribution**.

Apparently, the bytebuddy.net website can't qualify as "provided with distribution". 

4) The licenses change over time, and  http://bytebuddy.net does not distinguish between BB versions

and so on. 

> If the source web site is "other materials provided with the distribution", then showing the ASM license there should be sufficient (as you do on http://bytebuddy.net).

Please, be very careful if you suggest that to somebody else.

**Comment by ebruneton** [comment]
Maybe we can ask here: https://opensource.stackexchange.com/questions/7575/bsd-3-clause-where-to-place-license-for-binary-installation.

**Comment by vlsi** [comment]
@ebruneton , what is your point here?
Putting all the licenses under META-INF/LICENSE helps everybody, and it is trivial to do.

**Comment by ebruneton** [comment]
I'm not saying that the ByteBuddy jar should not include the ASM license in its meta-inf/, I'm saying that I'm not sure if this is sufficient to comply with the second clause of the license.

If not (and we can ask license experts to get an answer), maybe distributing ByteBuddy in a jar alone ( (but with ASM repackaged inside) is not compliant with the BSD license, and a zip should be distributed instead (containing the ByteBuddy jar, plus other materials, including the ASM license in a plain text file) [as you do in apache-jmeter-5.1.1.zip].

**Comment by vlsi** [comment]
> I'm saying that I'm not sure if this is sufficient to comply with the second clause of the license.

Ok, I see.

My point here is:
1) "jar without license" complicates life for everybody. asm.jar does not have license (yet?), and it complicates build for lots of libraries and application that bundle asm
2) BB uses need to somehow get all the licenses for the bundled software. It would be really nice if BB would just include the relevant bundled licenses
3) URLs are mortal, so they cannot be used as the only place for the license
4) It is sad that ASM developers refrain to include LICENSE to the asm.jar. So everybody ask them to include one.

> maybe distributing ByteBuddy in a jar alone ( (but with ASM repackaged inside) is not compliant with the BSD license

That might be very surprising since jar is probably 99.42% of all ByteBuddy distributions since it is extremely convenient to consume from Maven Central / JCenter / etc.

> as you do in apache-jmeter-5.1.1.zip

Apache JMeter is an application, so it is natural to include readme, docs, and things like that.
However, I would like to point out that `zip` file is binary, and it does not differ much from `jar`.
So putting bytebuddy.jar into bytebuddy.zip won't make a difference.

**Comment by ebruneton** [comment]
I'm not discussing about ASM itself here.

I just posted my question about ByteBuddy in https://opensource.stackexchange.com/questions/8676/bsd-3-clause-library-redistribution-in-a-java-library-as-a-jar-file.

**Comment by raphw** [comment]
I am not a lawyer but given that the license is so vague, I think that any form suffices. In such cases, a court of law would look into what the convention is and given that plugins such as the Maven Shade plugin even have this functionality built in, I would argue that this form is sufficient.

A jar file is just a zip file so given that jars are just specialized forms I cannot see how this would make a difference.

**Comment by latk** [comment]
I answered that Stack Exchange question. I think the approach that ByteBuddy is now using is sufficient, i.e. that the ASM license is included within the JAR's META-INF and that the ASM license is referenced in the documentation on the website.

It would also be possible to mention third party components within the NOTICE file, or to even include the entire ASM license in the NOTICE file. This would make it easier for downstream users to comply as they would only have to copy a single file. However, a `META-INF/license` directory for third party notices is fine as well, especially since this approach is customary in the Java ecosystem.

**Comment by raphw** [comment]
This was my understanding, too, at first where for example the Maven Shade plugin [preserves multiple `NOTICE` files](https://maven.apache.org/plugins/maven-shade-plugin/examples/resource-transformers.html#ApacheNoticeResourceTransformer) and it [collects the original licenses](https://maven.apache.org/plugins/maven-shade-plugin/examples/resource-transformers.html#ApacheLicenseResourceTransformer).

I did however not understand how this goes together with the [official documentation](http://www.apache.org/dev/licensing-howto.html) where it mentions that:

> Under normal circumstances, there is no need to modify NOTICE.

especially for the example of the *BSD 3-clause* license. Therefore, I followed the official guide where I did not touch the `NOTICE` file but only added a pointer to `LICENSE` while including the enforced license.
